### PR TITLE
Add support for establishing client channel using existing FD

### PR DESCRIPTION
**Pull request by xyzzyz**
This adds support for using existing FD as a client channel, as in https://github.com/grpc/grpc/issues/3250#. 

It seems to work correctly, but I'd like to hear some feedback with regard to whether I'm doing it in a correct way, and if there is something I should be doing, but I'm not. Also, I made pretty nasty hack for client to send the :authority header, so I'd like to ask what's the proper way of doing it. As I have no idea if the design is correct, I haven written any tests, and will do it once we have a shared understand of how the implementation and the tests look like.

Thanks,
Adam

### CODE REVIEW

File: src/core/ext/transport/chttp2/client/insecure/channel_create.c [hunk 1]
```diff
@@ -219,13 +222,115 @@ grpc_channel *grpc_insecure_channel_create(const char *target,
...
+     client_fd_channel_factory_create_channel};
+
+grpc_channel *grpc_insecure_channel_create_from_fd(
+    const char *target, int fd, const grpc_channel_args *args) {
+  grpc_exec_ctx exec_ctx = GRPC_EXEC_CTX_INIT;
+  GRPC_API_TRACE(
+      "grpc_insecure_channel_create(target=%p, fd=%d, args=%p)", 3,
+      (target, fd, args));
+
+  client_fd_channel_factory *f = gpr_malloc(sizeof(*f));
```
Comment (ctiller) [inline src/core/ext/transport/chttp2/client/insecure/channel_create.c]: You shouldn't need a channel factory here. And below, the channel type should be DIRECT.

Take a look at https://github.com/grpc/grpc/pull/6214/files#diff-9e0e02c477ce6a023a560178c2d1b079 - this is creating a simple channel around a custom transport. The code in grpc_insecure_channel_create_from_fd() should be directly creating the endpoint, creating the transport, and then just calling returning the result of grpc_channel_create.

I'd suggest taking the linked PR, making a new channel_create_from_fd.c from cronet_channel_create.c, and then copying in the relevant incantations from here.

Comment (ctiller) [inline src/core/ext/transport/chttp2/client/insecure/channel_create.c]: (you won't need to deal with the channel factory business at all then, and it should be fairly direct simple code)

**Comment by xyzzyz** [comment]
How about now? The channel is created directly, without any factory.

The question about authority is still open -- HTTP2 require :authority to be taken from the URL of the server, but we have no URL here. 

Also, what method of unit testing this would you recommend?

**Comment by xyzzyz** [comment]
Ping?

**Comment by ctiller** [comment]
Hey - my response got eaten somewhere. Lgtm so far. Probably best to add an end2end test fixture around this to get good coverage of any unusual paths this might introduce. Look at test/core/end2end/fixtures/h2_socketpair.c for instance (and test/core/end2end/gen_build_yaml.py to configure it in the build)

**Comment by xyzzyz** [comment]
Thanks for the response! 

What about the HTTP :authority header?

**Comment by xyzzyz** [comment]
Ping about the :authority.

Also, I have one more question -- with the API as it is now, this should probably be POSIX-only functionality. What's the proper way of making it available only on POSIX? 

Alternatively, what's the way to make it multiplatform? Making it take grpc_endpoint* instead of int fd probably is not a good idea, as we would have to expose a way of creating these.

**Comment by xyzzyz** [comment]
Huh, seems like I messed up this pull request somehow. I don't even know what happened -- tons of upstream commits got into it. Let me create a new one.

**Pull request closed without merging**